Some minor fixes to set_permissions() in file_permissions.py

payerle opened the pull request:
The set_permissions() routine claims to prevent users from creating
world writable suid binaries.  However, it seems to only be checking
for/preventing group writable suid binaries.

This patch modifies the routine to check for both world and group
writable suid binaries, and complain appropriately.

(This is a resubmission of #15544 which seems to have slipped between the cracks, and since then I  deleted and recreated my fork of spack, so I am not sure how viable the former is.

I first encountered this when building singularity a while back and confused by the "world-writable" error message on a file which was only group writable.  

I see that since then, singularity package.py creates a script to add suid rather than go through this script --- that is a better way in my opinion.  But while the set_permissions routine still exists, it should correctly check both group and world write perms on suid binaries and report problems correctly.
)

payerle committed:
Some minor fixes to set_permissions() in file_permissions.py

The set_permissions() routine claims to prevent users from creating
world writable suid binaries.  However, it seems to only be checking
for/preventing group writable suid binaries.

This patch modifies the routine to check for both world and group
writable suid binaries, and complain appropriately.

tldahlgren commented:
@tgamblin @becker33 Wouldn't we want to allow group-writable suid binaries for service accounts?

(Note the patch coverage fails for lack of a test for precluding group-writable suid binaries.)

tldahlgren reviewed (commented):
LGTM except for my question about allowing group-writable suid binaries.

payerle commented:
Please note that the proposed patch is _not_ adding a check for group writable SUID files, but for world/other writable SUID files.

The existing devel branch of spack throws exception on group-writable SUID files, but with a message and comments claiming it is checking for world/other writable SUID files.  Thus disallowing group writable SUID is and has been the behavior of spack since file_permissions.py was added 13 months ago with #11337.  Currently, despite what the comments in the code state, the check is for group writable (IWGRP) SUID files *only*, and there is no check for world writable SUIDs.

This patch _corrects_ the comments/error message for the group-writable SUID check, and _adds_ a check for world/other writable SUID.  

As for test coverage, existing tests in permissions.py only test for group writable SUID, e.g.
test_chmod_rejects_group_writable_suid

tldahlgren commented:
> 
> This patch _corrects_ the comments/error message for the group-writable SUID check, and _adds_ a check for world/other writable SUID.

Agreed!  The point I was trying to make was that I was questioning whether the original code was correct to begin with, especially given the inconsistency between the comment and the code.

> As for test coverage, existing tests in permissions.py only test for group writable SUID, e.g.
> test_chmod_rejects_group_writable_suid

I was referring to what was being reported in the code coverage results; namely, the exception for the group writable path is not being raised in the tests.

payerle commented:
OK, I will let those more knowledgeable about spack decide about the wisdom of blocking group-writable SUIDs --- I am relatively new to spack.  (My 2 cents: I like the approach taken by singularity as of  #16088 for non-suid builds --- do install as non-root but generate a small script sysadmin can vet and run as root to set suid stuff.)

I will look into adding unit test for world writable case.

payerle committed:
permissions.py: Add test to check blocks world writable SUID files

The original test_chmod_rejects_group_writable_suid tested
that the set_permissions() function in
lib/spack/spack/util/file_permissions.py
would raise an exception if changed permission on a file with
both SUID and SGID plus sticky bits is chmod-ed to g+rwx and o+rwx.

I have modified so that more narrowly tests a file with SUID
(and no SGID or sticky bit) set is chmod-ed to g+w.

I have added a second test test_chmod_rejects_world_writable_suid
that checks that exception is raised if an SUID file is chmod-ed
to o+w

payerle committed:
file_permissions.py: Raise exception when try to make sgid file world writable

Updated set_permissions() in file_permissions.py to also raise
an exception if try to make an SGID file world writable.  And
added corresponding unit test as well.

payerle commented:
Apparently the test was a bit overly-broad as well.  I have narrowed the scope of the existing test_chmod_rejects_group_writable_suid test to basically only do what it claims to do in the name, and added a complimentary test_chmod_rejects_world_writable_suid

I also added a patch to check for attempts to make an SGID file world writable.

tldahlgren commented:
> Apparently the test was a bit overly-broad as well. I have narrowed the scope of the existing test_chmod_rejects_group_writable_suid test to basically only do what it claims to do in the name, and added a complimentary test_chmod_rejects_world_writable_suid

Great!  That change addressed the uncovered lines.  Thanks.

payerle commented:
May I suggest that we defer on the issue re permitting group writable SUID files being produced by spack and approve the request as is?  

I am not sure there is really a good use case for allowing group writable SUID files:
1) they have been disabled by set_permissions() for over a year now
2) even if there is a valid use case for such, it still potentially opens security issues in the general case, and I would recommend adding a --dont-blame-spack or similar flag to spack to demote the error alerts in set_permissions to just warnings; thereby protecting the novice user while allowing those needing the functionality to use it.  (Unfortunately, I don't think either my python coding or familiarity with the workings of spack to be up to that)

becker33 commented on lib/spack/spack/test/permissions.py:
```diff
@@ -27,9 +27,32 @@ def test_chmod_real_entries_ignores_suid_sgid(tmpdir):
 def test_chmod_rejects_group_writable_suid(tmpdir):
     path = str(tmpdir.join('file').ensure())
-    mode = stat.S_ISUID | stat.S_ISGID | stat.S_ISVTX
+    print(type(tmpdir))
```
This looks like vestigial debugging code, please remove it.

becker33 commented on lib/spack/spack/test/permissions.py:
```diff
@@ -27,9 +27,32 @@ def test_chmod_real_entries_ignores_suid_sgid(tmpdir):
+
+def test_chmod_rejects_world_writable_suid(tmpdir):
+    path = str(tmpdir.join('file').ensure())
+    print(type(tmpdir))
```
Same here.

becker33 commented on lib/spack/spack/test/permissions.py:
```diff
@@ -27,9 +27,32 @@ def test_chmod_real_entries_ignores_suid_sgid(tmpdir):
+
+def test_chmod_rejects_world_writable_sgid(tmpdir):
+    path = str(tmpdir.join('file').ensure())
+    print(type(tmpdir))
```
Same here.

becker33 reviewed (changes requested):
@payerle minor requests, mostly LGTM

payerle committed:
Remove debugging prints from permissions.py
